### ISSUE TITLE
useIfMounted hook proposal

### ISSUE THREAD

**Issue by Velenir** [gnosis/dex-react#132]
We use `mounted = useRef()` a lot. Even if you ignore duplicated code, it's a pain to write it every time

Here's a proposal for a hook that could take care of the whole logic.

```ts
const useIfMounted = () => {
  const mounted = useRef(false);

  useEffect(() => {
    // sets on mount
    mounted.current = true;
    return () => {
      // resets on unmount
      mounted.current = false;
    };
  }, []);

  return useCallback<IFElseFunc>(
    <T extends AnyFunc, U extends AnyFunc>(
      ifMountedFunc: T | null | undefined,
      ifNotMountedFunc?: U
    ): ReturnType<T | U> | void => {
      if (mounted.current && ifMountedFunc) return ifMountedFunc();
      if (!mounted.current && ifNotMountedFunc) return ifNotMountedFunc();
    },
    []
  );
};

type AnyFunc = (...args: any[]) => any;

interface IFElseFunc {
  <T extends AnyFunc>(ifMountedFunc: T): ReturnType<T> | void;
  <T extends AnyFunc, U extends AnyFunc>(
    ifMountedFunc: T,
    ifNotMountedFunc: U
  ): ReturnType<T | U>;
  <U extends AnyFunc>(
    ifMountedFunc: undefined | null,
    ifNotMountedFunc: U
  ): void | ReturnType<U>;
}
```
I know types look crazy :smile: , but that's what you have to do for type inference
You can play with the hook here: https://codesandbox.io/s/useifmounted-hook-utees

**Comment by alfetopito** [comment]
Totally support the idea :+1:

**Comment by Velenir** [comment]
Proposal #2
titled : _Why didn't I think of this first?_

```ts
interface UseSafeState {
  <S = undefined>(initialState?: S | (() => S)): [
    S,
    Dispatch<SetStateAction<S>>
  ];
}

const useSafeState: UseSafeState = initialState => {
  const [state, setState] = useState(initialState);

  // can drop useRef, because we only need one closure
  // that is created at the start
  // and preserved thanks to deps = []
  let mounted = false;

  useEffect(() => {
    // the `mounted` from the very first render
    // is saved in this closure's outer scope
    // eslint-disable-next-line react-hooks/exhaustive-deps
    mounted = true;
    return () => {
      // resets on unmount
      mounted = false;
    };
  }, []);

  const setSafeState = useCallback(newState => {
    // this is the same `mounted` as in useEffect
    // deps are the same ([]) so closure's outer scope is the same
    if (mounted) setState(newState);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return [state, setSafeState];
};

// Usage:drop-in replacement for useState
...
const [count, setCount] = useSafeState()
// viva simplicity
```

https://codesandbox.io/s/usesafestate-hook-vkwnx

**Comment by alfetopito** [comment]
The provided interface for the second option seems more user friendly. I vote for number 2

**Comment by alfetopito** [comment]
Proposal 2 was implemented on #155
